### ISSUE TITLE
Confusion Matrix Issue After Training

### ISSUE THREAD

**Issue by smtalds-knt** [ultralytics/yolov5#13613]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I conducted a 300-epoch training using the BDD100K dataset. My test results are shown below. However, there seems to be a problem with the confusion matrix — the predictions appear incorrect or inconsistent.

Note:
Before the training phase, I followed the instructions in this GitHub issue comment:
https://github.com/Xilinx/Vitis-AI/issues/1252#issuecomment-1608431434
Specifically, I replaced the SiLU activation layer as recommended for compatibility with Vitis AI.

Could this modification be related to the issue with the confusion matrix?

![Image](https://github.com/user-attachments/assets/14220690-e3cf-42a6-a79f-8a9391a56b39)

![Image](https://github.com/user-attachments/assets/30cc3ffa-52fa-40b8-a80e-66e85976e0ec)

![Image](https://github.com/user-attachments/assets/5b3a8354-ba3b-4ecc-b831-8788f28af20d)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @smtalds-knt, thank you for your interest in YOLOv5 🚀! Please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).

If this is a 🐛 Bug Report, please provide a **minimum reproducible example** (MRE) that clearly demonstrates the confusion matrix issue, including:
- The exact commands used to train and evaluate
- Any code changes (such as your activation layer modification)
- Sample data or screenshots showing the problem

If this is a custom training ❓ Question, please share as much detail as possible, such as dataset examples, training logs, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```

## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your issue and assist you soon! 🔎

**Comment by pderrenger** [comment]
Yes, replacing the SiLU activation function can definitely impact your model's performance and confusion matrix results. SiLU is the default activation in YOLOv5 for good reason - it provides smooth gradients and typically better convergence compared to alternatives like ReLU.

To verify if the activation change is causing the confusion matrix inconsistencies, I'd recommend training a baseline model with the original SiLU activations on the same BDD100K dataset subset and comparing the [confusion matrix results](https://www.ultralytics.com/glossary/confusion-matrix). Also ensure you're using the latest YOLOv5 version as there have been improvements to validation metrics calculation.

The Vitis AI compatibility modification may be necessary for your deployment target, but it's likely contributing to the performance differences you're observing.

**Comment by smtalds-knt** [comment]
Hi @pderrenger ,

Thank you for your response.  
I am now able to get correct detections with high confidence scores from the FPGA devices — it's working well for me.

I had to avoid using `SiLU`, as it is not supported by Vitis AI.

Best regards,  
Samet

**Comment by pderrenger** [comment]
Great to hear that your FPGA deployment is working well with high confidence detections! You're absolutely right that avoiding SiLU was necessary for Vitis AI compatibility - this is a known trade-off when deploying to FPGA targets where certain activation functions aren't supported.

It sounds like you've successfully navigated the compatibility requirements while maintaining good model performance for your specific deployment scenario.
